### PR TITLE
Add optimization blacklist as second arg to optimizeForMobile method.

### PR DESCRIPTION
**Pull request by xcheng16**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37462 Add optimization blacklist as second arg to optimizeForMobile method.**

Summary:
Instead of running all the optimization pass in optimizeForMobile method,
introducing a whitelist optimizer dictionary as second param in the method,
when it is not passed during calling, the method will run all the optimization
passes, otherwise the method will read the dict and only run the pass with
value of True.

Test Plan:
python test/test_mobile_optimizer.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22096029](https://our.internmc.facebook.com/intern/diff/D22096029)

### CODE REVIEW

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 1]
```diff
@@ -296,13 +296,28 @@ void FoldPrePackingOps(script::Module& m) {
...
+    const script::Module& m,
+    const std::map<MobileOptimizerType, bool>& whitelist_optimizers) {
   auto cloned_module = m.clone();
   cloned_module.eval();
-  cloned_module = FoldConvBatchNorm2d(cloned_module);
-  insertPrePackedOps(cloned_module);
-  cloned_module = freeze_module(cloned_module);
-  FoldPrePackingOps(cloned_module);
+
+  auto bn_it =
```
Comment (AshkanAliabadi) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:305]: I would personally mark these as const auto (constness is not implied otherwise here) to have the compiler help me in situations where I accidentally modify the variable, like for instance `if (bn_it = whitelist_optimizers.end())`.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 2]
```diff
@@ -296,13 +296,28 @@ void FoldPrePackingOps(script::Module& m) {
...
+
+  auto prepack_it =
+      whitelist_optimizers.find(MobileOptimizerType::INSERT_FOLD_PREPACK_OPS);
+  if (prepack_it == whitelist_optimizers.end() ||
+      (prepack_it != whitelist_optimizers.end() && prepack_it->second)) {
+    insertPrePackedOps(cloned_module);
+    cloned_module = freeze_module(cloned_module);
+    FoldPrePackingOps(cloned_module);
+  }
+
```
Comment (AshkanAliabadi) [inline torch/csrc/jit/passes/xnnpack_rewrite.cpp:320]: This pattern is fine for two or three use cases but when and if the list of options grow we want to switch to something like this:

```
// Psudo code - most probably won't compile without modifications.
struct OpCode2Op {
  MobileOptimizerType opcdoe;
  std::function< script::Module (const script::Module&) > op;
};

const std::vector<OpCode2Op> opcodes2ops = {
  // OP 1
  {
     MobileOptimizerType::FOLD_CONV_BATCH_NORM,
     [](const script::Module& module) {
           return FoldConvBatchNorm2d(cloned_module);
     },
  },
  // OP 2
  {
    ...
  },
  // OP 3 and so on
  ...
};

auto cloned_module = m.clone();
cloned_module.eval();

foreach element in opcodes2ops:
  if element.opcode not in blacklist or is it is but not disabled:
     cloned_module = element->op(cloned_module);
```

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp [hunk 3]
```diff
@@ -296,13 +296,28 @@ void FoldPrePackingOps(script::Module& m) {
   PrePackingOpsFolder(m, filter_fn, "prepack_folding");
 }
 
-c10::optional<script::Module> optimizeForMobile(const script::Module& m) {
+script::Module optimizeForMobile(
+    const script::Module& m,
+    const std::map<MobileOptimizerType, bool>& whitelist_optimizers) {
```
[the remark this answers is not in the document]
Comment (AshkanAliabadi) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:301]: Also do we need a bool?  It could as well be const std::vector<MobileOptimizerType> & blacklist.  If it's in the blacklist, then skip, otherwise apply.

**Review by AshkanAliabadi — COMMENTED** [review]
Thanks a lot for working on this Xingying.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 2 above)
Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:320]: I like this suggestion. We can have a `const static` instance of OpCode2Op, or call it OptimizationTypes as Ashkan showed?
```
struct OptimizationType {
  MobileOptimizerType type;
  std::function< script::Module (const script::Module&) > apply;
};
```
Also we may need to run optimization in particular order. Meaning if you run batch norm fused after prepacking to insert xnnpack conv/linear then bn folding wont work. So we need a way to capture that as well. std::vector kind of works for it, but it does not capture the intent.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:301]: Maybe `unordered_set` if possible instead of map?

File: test/test_mobile_optimizer.py [hunk 4]
```diff
@@ -80,6 +94,41 @@ def forward(self, x):
...
+        whitelist_optimizer_dict_with_bn_fold = {
+            MobileOptimizerType.FOLD_CONV_BATCH_NORM: True,
+            MobileOptimizerType.INSERT_FOLD_PREPACK_OPS: False
+        }
+        bn_fold_scripted_module = optimize_for_mobile(bn_scripted_module, whitelist_optimizer_dict_with_bn_fold)
+        self.assertEqual(len(torch.jit.export_opnames(bn_fold_scripted_module)), 1)
+        FileCheck().check_count("prim::CallMethod[name=\"forward\"]", 1, exactly=True) \
+                   .run(str(get_forward_graph(bn_fold_scripted_module._c)))
+        bn_input = torch.rand(1, 1, 6, 6)
+        torch.testing.assert_allclose(bn_scripted_module(bn_input), bn_fold_scripted_module(bn_input), rtol=1e-2, atol=1e-3)
```
Comment (kimishpatel) [inline test/test_mobile_optimizer.py:130]: Can you add test case to check when both optimizations are turned off?

File: test/test_mobile_optimizer.py [hunk 5]
```diff
@@ -80,6 +94,41 @@ def forward(self, x):
...
+        FileCheck().check_count("prim::CallMethod[name=\"forward\"]", 2, exactly=True) \
+                   .run(str(get_forward(bn_scripted_module._c).graph))
+
+        whitelist_optimizer_dict_with_bn_fold = {
+            MobileOptimizerType.FOLD_CONV_BATCH_NORM: True,
+            MobileOptimizerType.INSERT_FOLD_PREPACK_OPS: False
+        }
+        bn_fold_scripted_module = optimize_for_mobile(bn_scripted_module, whitelist_optimizer_dict_with_bn_fold)
+        self.assertEqual(len(torch.jit.export_opnames(bn_fold_scripted_module)), 1)
+        FileCheck().check_count("prim::CallMethod[name=\"forward\"]", 1, exactly=True) \
```
Comment (kimishpatel) [inline test/test_mobile_optimizer.py:127]: It will be good to check that no prepacked ops are there. Plus not batch_norm.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Looks great. Left a few comments.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (xcheng16) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:301]: The initial idea was by default we run through all the optimization pass, but also let customer to customize it by passing a map where they can assign the value for each pass true/pass. Both of you are right, in that case we could have just created a blacklist vector to make it simpler.

File: torch/csrc/jit/passes/xnnpack_rewrite.cpp (hunk 2 above)
Comment (xcheng16) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:320]: Thanks for the suggestion, It looks good. Let me create a struct and attach a function for each pass. @kimishpatel Do we need to cover running optimization in particular order here?

Comment (kimishpatel) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:320]: Yes the order matters. Therefore I am not sure if doing thing is necessarily good:
```
foreach element in opcodes2ops:
  if element.opcode not in blacklist or is it is but not disabled:
     cloned_module = element->op(cloned_module);
```
@AshkanAliabadi, what do you suggest? I am of the opinion that we shouldn't just loop over the vector. It gets rid of the order information. I suggest we do something like:
```
assuming optimization_blacklist is a set
if (!optimization_blacklist.count(ConvBNFusion)) {
optimization_types.find(ConvBNFusion)->apply();
}
```
Or we can just keep it simple to be honest. 
In fact, while I like the idea, lets do it when we actually need to do it. Right now we dont have that many so we can keep it simple.
```
if (!optimization_blacklist.count(ConvBNFusion)) {
FoldConvBatchNorm2d(module)
}
```

Comment (xcheng16) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:320]: I think not looping the list and just checking if it is in blacklist before executing if order matters. @AshkanAliabadi are you good with it?

Comment (AshkanAliabadi) [reply torch/csrc/jit/passes/xnnpack_rewrite.cpp:320]: Yes definitely.

Generally speaking anything I say is just a suggestion so please feel free to ignore.  I believe the person writing the code or developing a feature has the best context and understanding of what's going on so please do as what you think makes best sense.  Thanks again.

File: torch/utils/mobile_optimizer.py [hunk 6]
```diff
@@ -4,26 +4,33 @@
...
     DROPOUT = 3
     BATCHNORM = 4
 
-def optimize_for_mobile(script_module):
+def optimize_for_mobile(script_module, optimization_blacklist: Set[MobileOptimizerType] = None):
     """
     Args:
         script_module: An instance of torch script module with type of ScriptModule
-
+        whitelist_optimizer_dict: A dictionary with key type of MobileOptimizerType and value type of bool.
```
Comment (kimishpatel) [inline torch/utils/mobile_optimizer.py:20]: Fix the comment? It is optimization_blacklist and not whitelist.

**Review by kimishpatel — APPROVED** [review]
Looks good. Please fix the comment before landing.
